Web Inspector: Console messages may be lost on main frame navigation depending on timing

patrickangle opened the pull request:
#### 77830d12bea6680bcd70a3f2c48fe58b6c24a74f
<pre>
Web Inspector: Console messages may be lost on main frame navigation depending on timing
<a href="https://bugs.webkit.org/show_bug.cgi?id=251659">https://bugs.webkit.org/show_bug.cgi?id=251659</a>
rdar://104303833

Reviewed by Devin Rousso.

The existing heuristic to determine if console messages are being cleared as the result of a main frame navigation made
the assumption that console messages won&apos;t be received between the call to clear messages and the next tick in Web
Inspector. This isn&apos;t a safe assumption, and can be observed when logging immediately after a navigation on some
computers/under certain types of loads. Instead of inferring why the console is being cleared, introduce a `reason`
parameter that tells us if the request to clear is the result of a main frame navigation, removing any doubt. For
compatibility with older remote ends, we keep the existing heuristic, since its better than nothing.

* LayoutTests/inspector/console/messagesCleared-expected.txt:
* LayoutTests/inspector/console/messagesCleared.html:
- Ensure that a new session was signalled due to a refresh when the console is cleared.

* Source/JavaScriptCore/inspector/agents/InspectorConsoleAgent.cpp:
(Inspector::InspectorConsoleAgent::clearMessages):
(Inspector::InspectorConsoleAgent::mainFrameNavigated):
(Inspector::InspectorConsoleAgent::clearMessages):
(Inspector::InspectorConsoleAgent::reset): Deleted.
- Provide a reason for the clearing of the console.

* Source/JavaScriptCore/inspector/agents/InspectorConsoleAgent.h:
* Source/JavaScriptCore/inspector/protocol/Console.json:
* Source/WebCore/inspector/InspectorInstrumentation.cpp:
(WebCore::InspectorInstrumentation::didCommitLoadImpl):

* Source/WebInspectorUI/UserInterface/Controllers/ConsoleManager.js:
(WI.ConsoleManager):
(WI.ConsoleManager.prototype.messagesCleared):
- Use the reason to immediately determine how to handle clearing the console, instead of waiting until the next tick.

(WI.ConsoleManager.prototype._clearMessages):
- Extract the common logic that actually clears the console so we can reuse it for both the new reason-based path as
well as the compatibility path.

(WI.ConsoleManager.prototype._delayedMessagesCleared):
* Source/WebInspectorUI/UserInterface/Protocol/ConsoleObserver.js:
(WI.ConsoleObserver.prototype.messagesCleared):

Canonical link: <a href="https://commits.webkit.org/261512@main">https://commits.webkit.org/261512@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9e8b0d53eeaf90e313c0fd8399df905c40e40728

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/111774 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/20903 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/388 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/3553 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/120488 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/115833 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22262 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/11981 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/3298 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117539 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/16534 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99681 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/104771 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/98486 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/257 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/45507 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/100256 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13373 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/258 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/92324 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/11506 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/13869 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/9780 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/101496 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/19313 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52259 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/31549 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8018 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/15855 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/109535 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26954 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/JavaScriptCore/inspector/agents/InspectorConsoleAgent.cpp:
```diff
@@ -112,14 +105,25 @@ bool InspectorConsoleAgent::developerExtrasEnabled() const
     m_counts.clear();
 }
 
+void InspectorConsoleAgent::internalClearMessages(Inspector::Protocol::Console::ClearReason reason)
```
NIT: we could just call this `clearMessages` and have the different parameters be enough of a distinguisher :)

dcrousso commented on Source/JavaScriptCore/inspector/protocol/Console.json:
```diff
@@ -34,6 +34,12 @@
+        {
+            "id": "ClearReason",
+            "type": "string",
+            "enum": ["console-clear-api", "main-frame-navigation"],
```
we should just have this be `"console"` or `"console-api"` to match other uses of things like this elsewhere

NIT: i wonder if just `"navigation"` would be enough 🤔

patrickangle replied:
I don't think `navigation` is enough here. We only send this for main frame navigations, and that feels like an important distinction.

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/ConsoleManager.js:
```diff
@@ -245,6 +257,17 @@ WI.ConsoleManager = class ConsoleManager extends WI.Object
         this._lastMessageLevel = level;
     }
 
+    _internalClearMessages()
```
ditto with just `_clearMessages`

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/ConsoleManager.js:
```diff
@@ -299,3 +315,8 @@ WI.ConsoleManager.Event = {
 };
+
+WI.ConsoleManager.ClearReason = {
+    ConsoleClearAPI: "console-clear-api",
```
ditto

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/ConsoleManager.js:
```diff
@@ -176,24 +178,34 @@ WI.ConsoleManager = class ConsoleManager extends WI.Object
+            break;
+        case WI.ConsoleManager.ClearReason.MainFrameNavigation:
+            if (!WI.settings.clearLogOnNavigate.value)
+                break;
```
what about all the other stuff that's inside `_mainResourceDidChange`?

NIT: instead of a second `break` we could just
```
if (WI.settings.clearLogOnNavigate.value)
    this._clearMessages();
break;
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/ConsoleManager.js:
```diff
@@ -176,24 +178,34 @@ WI.ConsoleManager = class ConsoleManager extends WI.Object
             // This could be for a navigation or other reasons (like console.clear()).
             // If this was a reload, we may not want to dispatch WI.ConsoleManager.Event.Cleared.
             // To detect if this is a reload we wait a turn and check if there was a main resource change reload.
             setTimeout(this._delayedMessagesCleared.bind(this), 0);
```
do we need a `return;` after this?

dcrousso commented on Source/WebCore/inspector/InspectorInstrumentation.cpp:
```diff
@@ -752,12 +752,12 @@ void InspectorInstrumentation::didCommitLoadImpl(InstrumentingAgents& instrument
         if (auto* networkAgent = instrumentingAgents.enabledNetworkAgent())
             networkAgent->mainFrameNavigated(*loader);
 
+        if (auto* consoleAgent = instrumentingAgents.webConsoleAgent())
+            consoleAgent->mainFrameNavigated();
```
does the ordering of this actually matter?  if so we should either add a comment or figure out a way to put that into the code so it's not changed in the future (and write a test for it too)

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/ConsoleManager.js:
```diff
@@ -176,24 +176,39 @@ WI.ConsoleManager = class ConsoleManager extends WI.Object
-
-            this._lastMessageLevel = null;
+        // COMPATIBILITY (iOS 16.4, macOS 13.3): `Console.ClearReason` did not exist.
+        if (!InspectorBackend.Enum.Console?.ClearReason) {
```
NIT: could we also check for `if (!reason) {` instead?

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/ConsoleManager.js:
```diff
@@ -176,24 +176,39 @@ WI.ConsoleManager = class ConsoleManager extends WI.Object
+        case WI.ConsoleManager.ClearReason.ConsoleAPI:
+            this._clearMessages();
+            return;
+        case WI.ConsoleManager.ClearReason.MainFrameNavigation:
```
Style: missing newline

patrickangle committed:
Web Inspector: Console messages may be lost on main frame navigation depending on timing
https://bugs.webkit.org/show_bug.cgi?id=251659
rdar://104303833

Reviewed by Devin Rousso.

The existing heuristic to determine if console messages are being cleared as the result of a main frame navigation made
the assumption that console messages won't be received between the call to clear messages and the next tick in Web
Inspector. This isn't a safe assumption, and can be observed when logging immediately after a navigation on some
computers/under certain types of loads. Instead of inferring why the console is being cleared, introduce a `reason`
parameter that tells us if the request to clear is the result of a main frame navigation, removing any doubt. For
compatibility with older remote ends, we keep the existing heuristic, since its better than nothing.

* LayoutTests/inspector/console/messagesCleared-expected.txt:
* LayoutTests/inspector/console/messagesCleared.html:
- Ensure that a new session was signalled due to a refresh when the console is cleared.

* Source/JavaScriptCore/inspector/agents/InspectorConsoleAgent.cpp:
(Inspector::InspectorConsoleAgent::clearMessages):
(Inspector::InspectorConsoleAgent::mainFrameNavigated):
(Inspector::InspectorConsoleAgent::clearMessages):
(Inspector::InspectorConsoleAgent::reset): Deleted.
- Provide a reason for the clearing of the console.

* Source/JavaScriptCore/inspector/agents/InspectorConsoleAgent.h:
* Source/JavaScriptCore/inspector/protocol/Console.json:
* Source/WebCore/inspector/InspectorInstrumentation.cpp:
(WebCore::InspectorInstrumentation::didCommitLoadImpl):

* Source/WebInspectorUI/UserInterface/Controllers/ConsoleManager.js:
(WI.ConsoleManager):
(WI.ConsoleManager.prototype.messagesCleared):
- Use the reason to immediately determine how to handle clearing the console, instead of waiting until the next tick.

(WI.ConsoleManager.prototype._clearMessages):
- Extract the common logic that actually clears the console so we can reuse it for both the new reason-based path as
well as the compatibility path.

(WI.ConsoleManager.prototype._delayedMessagesCleared):
* Source/WebInspectorUI/UserInterface/Protocol/ConsoleObserver.js:
(WI.ConsoleObserver.prototype.messagesCleared):

Canonical link: https://commits.webkit.org/261512@main
